### PR TITLE
Add a pangeo-forge runner image

### PR DESCRIPTION
**Pull request by yuvipanda**
Adds a `pangeo/forge` image containing environment for pangeo-forge runs,
inherited off pangeo-notebook with the following additions:

1. `apache-beam[gcp]` installed to allow for [Apache Beam](https://beam.apache.org/)
    execution, as pangeo-forge is moving towards this.
2. Puts the `boot` program that apache-beam needs into the appropriate place
3. A custom `start` script that launches either the apache-beam boot program or acts 
    as a no-op based on weather apache-beam is calling the container or some other
    program (such as binder)
4. Installs the `pangeo-forge-recipes` package into forge, rather than pangeo-notebook,
    so users can test it in binder as the 'playground' where needed.

I've also added test cases to make sure these features continue to work. In addition,
the test harnass wasn't actually testing the `Dockerfiles` in the other images - it was
writing a custom `Dockerfile` and testing *that*. This only tested the `ONBUILD` 
triggers from the base image, but not any customizations. This has been fixed.,

I'm installing `apache-beam[gcp]` via `pip` rather than `apache-beam-with-gcp` from
`conda-forge` as the latter seems to be broken - with it, beam was complaining about
not being able to find the libraries needed to talk to Google Cloud Store. This probably
needs to be fixed in [conda feedstock](https://github.com/conda-forge/apache-beam-feedstock)
somehow.

### CODE REVIEW

**Comment by yuvipanda** [comment]
Hold on, just testing and working with Charles at SciPy sprints

**Comment by cisaacstern** [comment]
To test this image with a real Pangeo Forge recipe, we did the following:

1. Built this image locally (which is a self-contained version of the image proposed in this PR)

    ```Dockerfile
    FROM golang:1.18-bullseye

    ENV VERSION=2.40.0
    RUN wget https://github.com/apache/beam/archive/refs/tags/v${VERSION}.tar.gz -O beam.tgz
    RUN tar xf beam.tgz
    
    RUN cd beam-${VERSION}/sdks/python/container && \
        CGO_ENABLED=0 go build -o /beam-boot
    
    FROM pangeo/pangeo-notebook:2022.07.13
    
    # FIXME: Pin version
    RUN pip install apache-beam[gcp]==2.40.0 --no-cache
    
    COPY --from=0 /beam-boot /usr/local/bin/boot
    
    ENTRYPOINT ["/usr/local/bin/boot"]
    ```
2. Appended this code (adapted from [this example](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/wordcount.py)) to a recipe which we know works on Dataflow

    ```python
    # recipe.py
    # The following recipe is copy-and-pasted above this code:
    # https://github.com/pangeo-forge/gpcp-feedstock/blob/4794d968c7d455bd6eef857864a5821a56bf4817/feedstock/recipe.py

    import argparse
    import logging
    
    import apache_beam as beam
    from apache_beam.options.pipeline_options import PipelineOptions
    from apache_beam.options.pipeline_options import SetupOptions
    
    from fsspec.implementations.local import LocalFileSystem
    from pangeo_forge_recipes.recipes import setup_logging
    from pangeo_forge_recipes.storage import FSSpecTarget
    
    setup_logging()
    local_fs = LocalFileSystem()
    recipe.storage_config.target = FSSpecTarget(local_fs, "test.zarr")
    recipe.cache_inputs = False
    chained_transform = recipe.copy_pruned().to_beam()
    
    parser = argparse.ArgumentParser()
    _, pipeline_args = parser.parse_known_args()
    pipeline_options = PipelineOptions(pipeline_args)
    pipeline_options.view_as(SetupOptions).save_main_session = True
    
    with beam.Pipeline(options=pipeline_options) as p:
        p | chained_transform
    ```

3. Run this `recipe.py`, with the following command [adapted from here](https://beam.apache.org/documentation/runtime/environments/#running-pipelines) (where `IMAGE_URL` is the image built in step 1 above)

    ```console
    export IMAGE="my-repo/beam_python_sdk_custom"
    export TAG="X.Y.Z"
    export IMAGE_URL = "${IMAGE}:${TAG}"
    
    python recipe.py \
    --runner=PortableRunner \
    --job_endpoint=embed \
    --environment_type="DOCKER" \
    --environment_config="${IMAGE_URL}"
    ```

4. As [described here](https://cloud.google.com/dataflow/docs/guides/using-custom-containers#basic_testing_with_portablerunner), the resulting zarr store will be on disk _inside the stopped container_ used for this run. To extract it, we call

    ```console
    docker cp $STOPPED_CONTAINER_ID:test.zarr test.zarr
    ```

5. `test.zarr` is now copied to our local filesystem, and we can open it with xarray

**Comment by cisaacstern** [comment]
TODOs:

- [ ] Work with @scottyhq to get this PR merged
- [ ] Get Scott's insight re: how we can bump the `pangeo-forge-recipes` for this `forge` image to generate new `forge` tags without disrupting the release cycle for other images in this repo
- [ ] Once released, use this image for the Pangeo Forge Sandbox here: https://github.com/pangeo-forge/sandbox/blob/binder/binder/Dockerfile
- [ ] Build on @ncclementi's suggestion to create a meta-package like https://github.com/coiled/coiled-runtime, (`pangeo-forge-runtime`?) which is a single conda-installable package that will bring in all of the same dependencies as this image, so that users have a conda-installable pathway for replicating the cloud worker environment.
- [ ] This meta-package could be used in the CI environments here https://github.com/pangeo-forge/pangeo-forge-recipes/tree/master/ci ?
- [ ] Upstream Yuvi's approach in this image to the Beam docs here: https://beam.apache.org/documentation/runtime/environments/#modify-existing-base-image

**Comment by TheNeuralBit** [comment]
:wave: hi @yuvipanda - I'm an Apache Beam maintainer. @cisaacstern brought this to my attention at the Beam Summit. I wanted to ask - is there something Beam could do that would make this easier? We definitely want it to be easy to build custom containers that include scientific libraries like this.

**Comment by yuvipanda** [comment]
Hello @TheNeuralBit! Excited to have you pop over here :)

What would really help is if *all* that is needed for an image to be runnable on beam is for it to have the apache-beam python package installed. I am one of the maintainers of [jupyterhub](https://github.com/jupyterhub/jupyterhub) where we support arbitrary images run on infrastructure, and the requirement is that the `jupyterhub` package is installed in the image and the `jupyterhub-singleuser` is a well-known command that is available in $PATH. This makes image building very flexible.

So in an ideal world, just installing `apache-beam` will provide the functionality that `boot` does now. This can be accomplished in two ways:

1. Ship the `boot` binary in the python package! This is actually quite doable and not a bad idea at all, and you can add this as part of your `setup.py`. Then you just need to ask for `apache-beam-boot` or similar to be in $PATH.
2. Rewrite the `boot` binary in python.

Either of these would be very very helpful!

**Comment by scottyhq** [comment]
> What would really help is if all that is needed for an image to be runnable on beam is for it to have the apache-beam python package installed 

If I understand the basic motivation for this PR it is to circumvent this limitation right? If so, I'd suggest opening an issue in this repository to revisit simplifying the `forge` configuration in the future

File: .github/workflows/Test.yml
Comment (scottyhq) [inline .github/workflows/Test.yml:60]: oh nice, this is much better than the temporary dockerfile hack

**Comment by yuvipanda** [comment]
> If I understand the basic motivation for this PR it is to circumvent this limitation right? If so, I'd suggest opening an issue in this repository to revisit simplifying the `forge` configuration in the future

Pretty much. @TheNeuralBit do you think this would be welcome? Would love help in navigating that community and its dynamics!

**Comment by yuvipanda** [comment]
I don't want to block this PR on that though.

File: forge/Dockerfile
Comment (scottyhq) [inline forge/Dockerfile:15]:
```suggestion
FROM golang:1.18-bullseye as beamboot
```

**Comment by cisaacstern** [comment]
I've linked back to here again in https://github.com/apache/beam/issues/22349#issuecomment-1190654944, just so these two threads don't end up diverging. @TheNeuralBit, thanks so much for your engagement on this!

File: forge/Dockerfile
Comment (scottyhq) [inline forge/Dockerfile:39]:
```suggestion
COPY --from=beamboot /beam-boot /usr/local/bin/boot
```

i think this is clearer, feel free to use a different name though

File: forge/apt.txt
Comment (scottyhq) [inline forge/apt.txt:1]: maybe we should just install `git` from conda-forge in all the images...

Comment (yuvipanda) [reply forge/apt.txt:1]: @scottyhq i'd actually recommend against that unless we have a specific reason, due to https://github.com/conda-forge/git-feedstock/issues/113 :(

File: pangeo-notebook/environment.yml
Comment (scottyhq) [inline pangeo-notebook/environment.yml:62]: if removing these you can either run conda-lock locally to add a new lockfile, or just update all images with the `/condalock` in a comment in this PR.

File: forge/apt.txt
Comment (scottyhq) [reply forge/apt.txt:1]: ah good to know!

**Comment by scottyhq** [comment]
> how we can bump the pangeo-forge-recipes for this forge image to generate new forge tags without disrupting the release cycle for other images in this repo 

This would require some conditionals in the current CI scripts. On average it's been a new build ~once a week and a new tag every two weeks (for all images), so maybe we just stick with that for the time being and revisit this in the future.

**Comment by yuvipanda** [comment]
@scottyhq that's for the tagged (with dates) images right? Images with just commit hashes can be as frequent as necessary right? And the bakery can just use that when needed.

**Review by scottyhq — APPROVED** [review]
Looks good to me @yuvipanda and @cisaacstern! Thanks for all the comments and tests!

**Comment by scottyhq** [comment]
> that's for the tagged (with dates) images right? Images with just commit hashes can be as frequent as necessary right? 

Correct!

**Comment by yuvipanda** [comment]
@scottyhq looks like the pytorch image might be failing with some of the newer dependencies :( I'm going to try and revert the lock update across images and just do it for the forge.

File: forge/Dockerfile
Comment (alxmrs) [inline forge/Dockerfile:24]: Why not pull this from the Beam Python SDK directly? Are you sure you need to compile the go sources again? I'm not sure this is necessary.

Comment (yuvipanda) [reply forge/Dockerfile:24]: @abkfenris yeah, we could pull the go binary alone out of the beam sdk image (assuming that is what you are referring to) and I think that should be ok too. My initial reading of the beam dockerfile assumed that the python package was also installed into `/opt/apache/beam` and I didn't want to pull *that* in. But I guess I can copy just the binary out?

Do you know if it's statically built or links to local libc?

Comment (cisaacstern) [reply forge/Dockerfile:24]: Knowing @abkfenris, he'll probably be interested in this conversation, but you probably meant to tag @alxmrs. 😄

Comment (abkfenris) [reply forge/Dockerfile]: You mean I don't have to go down this rabbit hole right now? I mean I want to, but I really can't.

File: forge/Dockerfile
Comment (cisaacstern) [inline forge/Dockerfile:29]: Does it matter if you pull from `python3.8_sdk` vs. `python3.9_sdk`?

Comment (yuvipanda) [reply forge/Dockerfile:29]: @cisaacstern i'll switch it to 3.9 but it shouldn't matter as I'm picking up just the built go binary. Should modify docs too.

File: forge/Dockerfile
Comment (yuvipanda) [reply forge/Dockerfile]: @cisaacstern do you know what error you were getting?

File: forge/Dockerfile
Comment (alxmrs) [reply forge/Dockerfile:29]: Yes, it does matter: the python versions need to be consistent all the way through, or else you get Dataflow runtime errors. There is an open JIRA ticket in the beam project to address this, but I'm having a hard time finding it now.

Comment (yuvipanda) [reply forge/Dockerfile:29]: @alxmrs I'm only copying the `boot` go binary, not the built python package. The python package is being installed with `pip` into our conda environment directly here. I don't want to bring in any python dependencies by copying out of the beam image.

**Comment by yuvipanda** [comment]
Looks like pytorch's newest version is currently broken - https://github.com/pytorch/pytorch/issues/74473. I'm gonna temporarily pin it to unblock this PR

**Comment by yuvipanda** [comment]
@alxmrs ok, I'm now pulling just the go binary from the upstream docker image! It's no longer statically built, but I think in our specific case it's probably ok because we're still within the debian family. But I'm always wary of libc incompatibilities, so maybe worth making the build be static upstream too. 

Thank you so much for engaging with us here! If there is anything else that you think we could do here, please let us know?

**Comment by yuvipanda** [comment]
I opened https://github.com/pangeo-data/pangeo-docker-images/issues/358 to unpin pytorch when that's done

File: forge/conda-linux-64.lock
Comment (scottyhq) [inline forge/conda-linux-64.lock:37]: @yuvipanda i think this lock file is incorrect and actually corresponds to the ML notebook (cuda, tensorflow), i didn't see anything wrong with CI so maybe it got overwritten locally by accident

Comment (scottyhq) [reply forge/conda-linux-64.lock:37]:
```
cd forge
conda-lock lock --mamba -k explicit -f environment.yml -f ../pangeo-notebook/environment.yml -f ../base-notebook/environment.yml -p linux-64
```
doesnt have the ML libraries

Comment (yuvipanda) [reply forge/conda-linux-64.lock:37]: @scottyhq quite possible, i'm relocking!

File: forge/Dockerfile
Comment (scottyhq) [inline forge/Dockerfile:28]: Could be a place to use `COPY --link` if we eventually switch to dockerfile1.4 syntax (#357) https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#copy---link

File: forge/conda-linux-64.lock
Comment (yuvipanda) [reply forge/conda-linux-64.lock:37]: @scottyhq looks like the lockbot isn't locking forge yet, might need this to merge first. I'll try do it locally.

Comment (scottyhq) [reply forge/conda-linux-64.lock:37]: Yeah the `/condalock` command runs like `pull-request-target` so it doesn't see the new action, you have to run just the forge lock locally and then after merging it will work through CI

File: forge/Dockerfile
Comment (cisaacstern) [reply forge/Dockerfile:28]: TIL `COPY --link`. The docker knowledge on this repo/team are awesome 💪 !

**Comment by yuvipanda** [comment]
@scottyhq i've updated the lockfile (and made sure the packages.txt file contains pip components too), and merged it! Thanks a lot for review + approval!

**Pull request merged**
